### PR TITLE
Rewrite dumper in Go

### LINKED ISSUE #47: Dumper silently fails for ambiguous kinds with multiple groups
## Steps to Reproduce the Problem

If there are kinds available on the cluster with ambiguous names, then the dumper will just dump the objects for only one of them.

Example:
```
k get crd | grep data
databases.mssql.sql.crossplane.io                          2024-10-04T11:15:04Z
databases.mysql.sql.crossplane.io                          2024-10-04T11:15:04Z
databases.postgresql.sql.crossplane.io                     2024-10-04T11:15:04Z

k get crd | grep user
users.minio.crossplane.io                                  2024-10-04T11:15:00Z
users.mssql.sql.crossplane.io                              2024-10-04T11:15:04Z
users.mysql.sql.crossplane.io                              2024-10-04T11:15:04Z
```

All these CRDs have the database or user kind, but different groups.

```
k get databases.mssql.sql.crossplane.io 
No resources found

k get databases.postgresql.sql.crossplane.io 
NAME                          READY   SYNCED   AGE
beck-pg-tbc5b-test-database   True    True     25m

k get users.minio.crossplane.io 
NAME                 READY   SYNCED   EXTERNAL NAME        POLICIES             AGE
appcat-sla-reports   True    True     appcat-sla-reports   appcat-sla-reports   46m
beck-pg-tbc5b        True    True     beck-pg-tbc5b        beck-pg-tbc5b        40m

k get users.mssql.sql.crossplane.io 
No resources found
```

## Actual Behavior
It dumps all the objects for of each kind only for the first group it finds. The ordering is alphabetical.
```
objects-database.json
{
    "apiVersion": "v1",
    "items": [],
    "kind": "List",
    "metadata": {
        "resourceVersion": ""
    }
}

objects-users.json
{
    "apiVersion": "v1",
    "items": [
        {
            "apiVersion": "minio.crossplane.io/v1", # only minio users
            "kind": "User",
            "metadata": {
                "annotations": {
*SNIP*
```

## Expected Behavior
It should create lists for all the groups, not just one.

### PR DESCRIPTION
**Pull request by bastjan**
Differences:
- All APIs are fully qualified in both the options (`--must-exist=certificates.cert-manager.io`, `--ignore=deployment.apps`) and the output files (`objects-Certificate.cert-manager.io.json`). This makes it possible to distinguish between objects with the same kind but different groups. See https://github.com/projectsyn/k8s-object-dumper/issues/47.
- Resources without a list endpoint are ignored, do not cause an error, and don't need to be explicitly ignored.
- Ignore and must-exist options are now command line flags instead of files in `/usr/local/share`.

Fixes #47 and maybe #42.

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [x] Update the documentation.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog

### CODE REVIEW

File: CODE_OF_CONDUCT.md
Comment (simu) [inline CODE_OF_CONDUCT.md:1]: Arguably this should be preserved for repos in github.com/projectsyn

File: internal/pkg/discovery/discovery.go [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+// GetLogWriter returns the set batch size for listing objects or io.Discard as default.
+func (opts DiscoveryOptions) GetLogWriter() io.Writer {
+	if opts.LogWriter == nil {
+		return io.Discard
+	}
+	return opts.LogWriter
+}
+
+// DiscoverObjects discovers all objects in the cluster and calls the provided callback for each list of objects.
+// The callback can be called multiple times with the same
```
Comment (simu) [inline internal/pkg/discovery/discovery.go]: Incomplete comment, should this be

```suggestion
// The callback can be called multiple times with the same object
```

File: main.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+	flag.Var(ignoreResources, "ignore", "Resource to ignore during discovery. Regexp, anchored by default. Can be used multiple times.")
+
+	flag.Parse()
+
+	df := dumper.DumpToWriter(os.Stdout)
+	if dir != "" {
+		if err := os.MkdirAll(dir, 0755); err != nil {
+			fmt.Fprintf(os.Stderr, "failed to create directory %s: %v\n", dir, err)
+			os.Exit(1)
+		}
```
Comment (simu) [inline main.go:34]: This seems redundant since we do `os.MkdirAll(dir, 0755)` again in `dumper.NewDirDumper()`

**Review by simu — APPROVED** [review]
LGTM overall :rocket: 

The switch from files to command line flags for the must-exist and ignore resources configs will most likely require adjustments in https://github.com/projectsyn/component-cluster-backup

File: main.go (hunk 2 above)
Comment (bastjan) [reply main.go:34]: I added it as an early exit if some directory is not writeable.

File: internal/pkg/discovery/discovery.go (hunk 1 above)
Comment (bastjan) [reply internal/pkg/discovery/discovery.go]:
```
The callback can be called multiple times with the same object kind. Objects are unique in general, but might not be if the aggregate API-Server does not implement batching correctly.
```

Comment (bastjan) [reply internal/pkg/discovery/discovery.go]:
```
// DiscoverObjects discovers all objects in the cluster and calls the provided callback for each list of objects.
// The callback can be called multiple times with the same object kind. Objects are unique in general,
// but the callback should be able to handle duplicates. Some API-Servers do not implement list batching correctly and thus might introduce duplicates.
```

File: CODE_OF_CONDUCT.md
Comment (bastjan) [reply CODE_OF_CONDUCT.md:1]: Added, this got lost.

**Pull request merged**